File: Project 1/Main.java
/**
 * Nicholas Pham
 * CS 408: Programming Languages
 * Project 1: Set Data Structure
 * Description: Implement a Set data structure using ADTs and non ADTs. The Set operations that should surely be included are
 * 		union, intersection, and difference. Create a driver to test the operations and allow a user to use the operations.
 */

import java.util.ArrayList;
import java.util.Scanner;

public class Main {
	
	public static void main(String[] args) {
		Scanner kb = new Scanner(System.in);
		System.out.println("Which version of Set would you like to test?");
		System.out.println("1 - Non-ADT version");
		System.out.println("2 - ADT version");
		int input = kb.nextInt();
		if (input == 1) {
			nonADTSetTest();
		}
		else if (input == 2) {
			adtSetTest();
		}
		else {
			System.out.println("Input is not valid.");
		}
		
		
		kb.close();
	}
	
	/*********************************************************************/
	
	/**
	 * This method tests the non-ADT version of Set.
	 */
	public static void nonADTSetTest() {
		Scanner kb = new Scanner(System.in);
		ArrayList<Integer> nonADTSet1 = new ArrayList<>();
		ArrayList <Integer>nonADTSet2 = new ArrayList<>();

		System.out.println("Enter integers for Set 1 (-999 to stop): ");
		int element;
		element = kb.nextInt();
		while (element != -999) {
			if (!nonADTSet1.contains(element)) {
				nonADTSet1.add(element);
			}
			else {
				System.out.println("The integer \"" + element + "\" already exists in the set!");
			}
			element = kb.nextInt();
		}
		System.out.println("Set 1: " + nonADTSet1.toString());
		
		System.out.println("Enter integers for Set 2 (-999 to stop): ");
		element = kb.nextInt();
		while (element != -999) {
			if (!nonADTSet2.contains(element)) {
				nonADTSet2.add(element);
			}
			else {
				System.out.println("The integer \"" + element + "\" already exists in the set!");
			}
			element = kb.nextInt();
		}
		System.out.println("Set 2: " + nonADTSet2.toString());
		
		nonADTMenu();
		String input = kb.next();
		while (!input.equals("e")) {
			if (input.equals("u")) {
				ArrayList<Integer> nonADTUnion = union(nonADTSet1, nonADTSet2);
				System.out.println("Union of Set 1 and Set 2: " + nonADTUnion.toString());
				nonADTMenu();
				input = kb.next();
			}
			else if (input.equals("i")) {
				ArrayList<Integer> nonADTIntersection = intersection(nonADTSet1, nonADTSet2);
				System.out.println("Intersection of Set 1 and Set 2: " + nonADTIntersection.toString());
				nonADTMenu();
				input = kb.next();
			}
			else if (input.equals("d")) {
				ArrayList<Integer> nonADTDifference = difference(nonADTSet1, nonADTSet2);
				System.out.println("Difference of Set 1 and Set 2: " + nonADTDifference.toString());
				nonADTMenu();
				input = kb.next();
			}
			else {
				System.out.println("That is not a valid command.");
				nonADTMenu();
				input = kb.next();
			}
		}
		System.out.println("Exiting the program.");
	}

	/**
	 * This menu is for the non-ADT version of Set.
	 */
	public static void nonADTMenu() {
		System.out.println("What would you like to do with your sets?");
		System.out.println("\"u\" : perform union between the two sets");
		System.out.println("\"i\" : perform intersection between the two sets");
		System.out.println("\"d\" : perform difference between the two sets");
		System.out.println("\"e\" : exit this program");
	}
	
	/**
	 * This method is the non-ADT version of union.
	 */
	public static ArrayList<Integer> union(ArrayList<Integer> set1, ArrayList<Integer> set2) {
		ArrayList<Integer> output = new ArrayList<>();
		for (int i = 0; i < set1.size(); i++) {
			output.add(set1.get(i));
		}
		
		for (int i = 0; i < set2.size(); i++) {
			if (!set1.contains(set2.get(i))) {
				output.add(set2.get(i));
			}
			else {
				System.out.println("The element \"" + set2.get(i) + "\" already exists in the set!");
			}
		}
		return output;
	}
	
	/**
	 * This method is the non-ADT version of intersection.
	 */
	public static ArrayList<Integer> intersection(ArrayList<Integer> set1, ArrayList<Integer> set2) {
		ArrayList<Integer> output = new ArrayList<>();
		for (int element : set1) {
			if (set2.contains(element)) {
				output.add(element);
			}
		}
		return output;	
	}
	
	/**
	 * This method is the non-ADT version of difference.
	 */
	public static ArrayList<Integer> difference(ArrayList<Integer> set1, ArrayList<Integer> set2) {
		ArrayList<Integer> output = new ArrayList<>();
		output.addAll(set1);
		for (int i = 0; i < set2.size(); i++) {
			if (output.contains(set2.get(i))) {
				output.remove(set2.get(i));
				System.out.println("Removing: " + set2.get(i));
			}
			else {
				output.add(set2.get(i));
				System.out.println("Adding: " + set2.get(i));
			}
		}
		return output;
	}
	
	/*********************************************************************/
	
	/**
	 * This method tests the ADT version of Set.
	 */
	public static <T> void adtSetTest() {
		Scanner kb = new Scanner(System.in);
		// Ask for first Set
		SetADT<T> set1 = new SetADT<>();
		System.out.println("Enter values for Set 1: (\"exit\" to stop)");
		T input = (T)kb.next();
		while (!input.equals("exit")) {
			set1.addElement(input);
			input = (T)kb.next();
		}
		System.out.println("First set inputed: " + set1.toString());
		
		// Ask for second Set
		SetADT<T> set2 =  new SetADT<>();
		System.out.println("Enter values for Set 2: (\"exit\" to stop)");
		T input2 = (T)kb.next();
		while (!input2.equals("exit")) {
			set2.addElement(input2);
			input2 = (T)kb.next();
		}
		System.out.println("Second set inputed: " + set2.toString());
		
		// Menu
		printMenu();
		String menuInput = kb.next();
		while (!menuInput.equals("e")) {
			if (menuInput.equals("u")) {
				SetADT<T> set3 = set1.union(set2);
				System.out.println("Union of Set 1 and Set 2: " + set3.toString());
				printMenu();
				menuInput = kb.next();
			}
			else if (menuInput.equals("i")) {
				SetADT<T> set3 = set1.intersection(set2);
				System.out.println("Intersection of Set 1 and Set 2: " + set3.toString());
				printMenu();
				menuInput = kb.next();
			}
			else if (menuInput.equals("d")) {
			 	SetADT<T> set3 = set1.difference(set2);
				System.out.println("Difference of Set 1 and Set 2: " + set3.toString());
			 	printMenu();
			 	menuInput = kb.next();
			}
			else {
				System.out.println("That is not a valid command.");
				printMenu();
				menuInput = kb.next();
			}
		}
		System.out.println("Exiting the program.");
	}
	
	/**
	 * This method prints out the menu and allows for a cleaner interface.
	 */
	public static void printMenu() {
		System.out.println("What would you like to do with these sets now?");
		System.out.println("\"u\" : perform union between the two sets");
		System.out.println("\"i\" : perform intersection between the two sets");
		System.out.println("\"d\" : perform difference between the two sets");
		System.out.println("\"e\" : exit this program");
	}

}

File: Project 1/Main.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include "SetADT.cpp"
using namespace std;

// Non ADT Set
void nonADTSet();
// Union of two sets
vector<int> operator+(vector<int> lhs, vector<int> rhs);
// Intersection of two sets
vector<int> operator&(vector<int> set1, vector<int> set2);
// Difference of two sets
vector<int> operator-(vector<int> set1, vector<int> set2);
// Printing a set
void toString(vector<int>& v);
// Menu
void menu();
// ADT version
void adtSetTest();

int main() {

	cout << "Which version of Set would you like to test?" << endl;
	cout << "1 - Non-ADT version" << endl;
	cout << "2 - ADT version" << endl;

	int input;
	cin >> input;
	if (input == 1) {
		nonADTSet();
	}
	else if (input == 2) {
		adtSetTest();
	}
	else {
		cout << "Input is not valid." << endl;
	}

	return 0;
}

/**
 * This function performs the non-ADT version of Set
 */
void nonADTSet() {
	vector<int> nonADTSet1;
	vector<int> nonADTSet2;

	// Ask user to input the sets
	cout << "Enter integers for Set 1: (-999 to stop)" << endl;
	int element;
	cin >> element;
	while (element != -999) {
		if (find(nonADTSet1.begin(), nonADTSet1.end(), element) != nonADTSet1.end()) {
			cout << "The integer \"" << element << "\" already exists in the set!" << endl;
		}
		else {
			nonADTSet1.push_back(element);
		}
		cin >> element;
	}
	cout << "Set 1: ";
	toString(nonADTSet1);

	cout << "Enter integers for Set 2: (-999 to stop)" << endl;
	cin >> element;
	while (element != -999) {
		if (find(nonADTSet2.begin(), nonADTSet2.end(), element) != nonADTSet2.end()) {
			cout << "The integer \"" << element << "\" already exists in the set!" << endl;
		}
		else {
			nonADTSet2.push_back(element);
		}
		cin >> element;
	}
	cout << "Set 2: ";
	toString(nonADTSet2);

	// What to do with them
	menu();
	string input;
	cin >> input;
	while (input != "e") {
		if (input == "u") {
			vector<int> unionSet = nonADTSet1 + nonADTSet2;
			cout << "Union of Set 1 and Set 2: ";
			toString(unionSet);
			menu();
			cin >> input;
		}
		else if (input == "i") {
			vector<int> intersectionSet = nonADTSet1 & nonADTSet2;
			cout << "Intersection of Set 1 and Set 2: ";
			toString(intersectionSet);
			menu();
			cin >> input;
		}
		else if (input == "d") {
			vector<int> differenceSet = nonADTSet1 - nonADTSet2;
			cout << "Difference of Set 1 and Set 2: ";
			toString(differenceSet);
			menu();
			cin >> input;
		}
		else {
			cout << "That is not a valid command." << endl;
			menu();
			cin >> input;
		}
	}
	cout << "Exiting the program." << endl;
}

/**
 * This function performs union between two sets (vectors)
 */
vector<int> operator+(vector<int> set1, vector<int> set2) {
	vector<int> output;
	for (int i = 0; i < set1.size(); i++) {
		output.push_back(set1[i]);
	}
	for (int i = 0; i < set2.size(); i++) {
		if (find(output.begin(), output.end(), set2[i]) != output.end()) {
			cout << "The integer \"" << set2[i] << "\" already exists in the set!" << endl;
		}
		else {
			output.push_back(set2[i]);
		}
	}
	return output;
}

/**
 * This function performs intersection between two sets (vectors)
 */
vector<int> operator&(vector<int> set1, vector<int> set2) {
	vector<int> output;
	for (int i = 0; i < set1.size(); i++) {
		if (find(set2.begin(), set2.end(), set1[i]) != set2.end()) {
			output.push_back(set1[i]);
		}
	}
	return output;
}

/**
 * This function performs difference between two sets (vectors)
 */
vector<int> operator-(vector<int> set1, vector<int> set2) {
	vector<int> output;
	vector<int>::iterator it;
	for (int i = 0; i < set1.size(); i++) {
		output.push_back(set1[i]);
	}
	for (int i = 0; i < set2.size(); i++) {
		if (find(output.begin(), output.end(), set2[i]) != output.end()) {
			it = find(output.begin(), output.end(), set2[i]);
			output.erase(it);
		}
		else {
			output.push_back(set2[i]);
		}
	}
	return output;
}

/**
 * This function prints out the vector/set.
 */
void toString(vector<int>& v) {
	if (v.empty()) {
		cout << "[ ]" << endl;
	}
	cout << "[";
	for (int i = 0; i < v.size() - 1; i++) {
		cout << v[i] << ", ";
	}
	cout << v.back() << "]" << endl;
}

/**
 * This function prints the non-ADT menu.
 */
void menu() {
	cout << "What would you like to do with your sets?" << endl;
	cout << "\"u\" : perform union between the two sets" << endl;
	cout << "\"i\" : perform intersection between the two sets" << endl;
	cout << "\"d\" : perform difference between the two sets" << endl;
	cout << "\"e\" : exit this program" << endl;
}

void adtSetTest() {
	// Ask for first set
	SetADT set1;
	cout << "Enter values for Set 1: (\"-999\" to stop)" << endl;
	int input;
	cin >> input;
	while (input != -999) {
		set1.addElement(input);
		cin >> input;
	}
	cout << "First set inputed: ";
	set1.toString();

	// Ask for second set
	SetADT set2;
	cout << "Enter values for Set 2: (\"-999\" to stop)" << endl;
	int input2;
	cin >> input2;
	while (input2 != -999) {
		set2.addElement(input2);
		cin >> input2;
	}
	cout << "Second set inputed: ";
	set2.toString();

	// Menu
	menu();
	string menuInput;
	cin >> menuInput;
	while (menuInput != "e") {
		if (menuInput == "u") {
			SetADT set3 = set1 + set2;
			cout << "Union of Set 1 and Set 2: ";
			set3.toString();
			menu();
			cin >> menuInput;
		}
		else if (menuInput == "i") {
			SetADT set3 = set1 & set2;
			cout << "Intersection of Set 1 and Set 2: ";
			set3.toString();
			menu();
			cin >> menuInput;
		}
		else if (menuInput == "d") {
			SetADT set3 = set1 - set2;
			cout << "Difference of Set 1 and Set 2: ";
			set3.toString();
			menu();
			cin >> menuInput;
		}
		else {
			cout << "That is not a valid command." << endl;
			menu();
			cin >> menuInput;
		}
	}
	cout << "Exiting the program." << endl;
}
